test/assignment-5/CollateralizedVault.t.sol

```diff
@@ -0,0 +1,270 @@
+        weth = new WETH9();
+        priceFeedMock = new ChainlinkPriceFeedMock();
+        priceFeedMock.setPrice(500000000000000); // = 1/2000
+        vault = new CollateralizedVault(address(dai), address(weth), address(priceFeedMock));
+
+        setDaiBalance(address(vault), 10000 ether);
+        setWethBalance(USER, 10 ether);
+        vm.prank(USER);
+        weth.approve(address(vault), 10 ether);
+    }
+
+    function setDaiBalance(address dst, uint256 balance) public {
+        stdstore
+            .target(address(dai))
+            .sig(dai.balanceOf.selector)
+            .with_key(dst)
+            .depth(0)
+            .checked_write(balance);
+    }
+
+    function setWethBalance(address dst, uint256 balance) public {
+        stdstore
+            .target(address(weth))
+            .sig(weth.balanceOf.selector)
+            .with_key(dst)
+            .depth(0)
+            .checked_write(balance);
+    }
+}
+
+contract ZeroStateTest is ZeroState {
+
+    function testDeposit() public {
+        vm.prank(USER);
+        vault.deposit(3 ether);
+
+        // 3 WETH was transfered to the vault
+        assertEq(weth.balanceOf(USER), 7 ether);
+        assertEq(weth.balanceOf(address(vault)), 3 ether);
+
+        // User collateral is recorded
+        assertEq(vault.deposits(USER), 3 ether);
+    }
+
+    function testDepositMultipleTimesAccumulatesCollateral() public {
+        vm.startPrank(USER);
+
+        vault.deposit(1 ether);
+        vault.deposit(1 ether);
+
+        assertEq(weth.balanceOf(address(vault)), 2 ether);
+        assertEq(vault.deposits(USER), 2 ether);
+    }
+
+    function testScaleInteger() public {
+        // scale 12.345 from 6 decimals to 18
+        uint256 from = (12345 * 1e6) / 1e3;
+        uint256 to = vault.scaleInteger(from, 6, 18);
+
+        assertEq(to, (12345 * 1e18) / 1e3);
+
+        // test reverse
+        assertEq(vault.scaleInteger(to, 18, 6), from);
+    }
+}
+
+abstract contract DepositedCollateralState is ZeroState {
+    function setUp() public virtual override {
+        super.setUp();
+
+        vm.prank(USER);
+        vault.deposit(3 ether);
+
+        vm.prank(USER);
+        dai.approve(address(vault), type(uint256).max);
+    }
+}
+
+contract DepositedCollateralStateTest is DepositedCollateralState {
+    function testBorrow() public {
+        vm.prank(USER);
+        vault.borrow(6000 * 1e18);
+
+        // 6000 DAI was transfered to the USER
+        assertEqDecimal(dai.balanceOf(USER), 6000 * 1e18, 18);
+
+        // User debt is recorded
+        assertEq(vault.borrows(USER), 6000 * 1e18);
+    }
+
+    function testBorrowMultipleTimesAccumulatesDebt() public {
+        vm.startPrank(USER);
+
+        vault.borrow(5 * 1e18);
+        vault.borrow(10 * 1e18);
+
+        assertEqDecimal(dai.balanceOf(USER), 15 * 1e18, 18);
+        assertEq(vault.borrows(USER), 15 * 1e18);
+    }
+
+    function testRevertsWhenTryingToBorrowTooMuch() public {
```

Comment:
alberto always tells us to put our negative assertion tests at teh beginning of each state